File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/model/IceCandidate.kt
package com.instrumentisto.medea_flutter_webrtc.model

import org.webrtc.IceCandidate as WIceCandidate

/**
 * Representation of an [org.webrtc.IceCandidate].
 *
 * @property sdpMid `mid` of this [IceCandidate].
 * @property sdpMLineIndex `sdpMLineIndex` of this [IceCandidate].
 * @property sdp SDP of this [IceCandidate].
 */
data class IceCandidate(val sdpMid: String, val sdpMLineIndex: Int, val sdp: String) {
  companion object {
    /**
     * Creates a new [IceCandidate] object based on the method call received from the Flutter side.
     *
     * @return [IceCandidate] created from the provided [Map].
     */
    fun fromMap(map: Map<String, Any>): IceCandidate {
      return IceCandidate(
          map["sdpMid"] as String, map["sdpMLineIndex"] as Int, map["candidate"] as String)
    }

    /**
     * Converts the provided [org.webrtc.IceCandidate] into an [IceCandidate].
     *
     * @return [IceCandidate] created based on the provided [org.webrtc.IceCandidate].
     */
    fun fromWebRtc(from: WIceCandidate): IceCandidate =
        IceCandidate(from.sdpMid, from.sdpMLineIndex, from.sdp)
  }

  /**
   * Converts this [IceCandidate] to an [org.webrtc.IceCandidate].
   *
   * @return [org.webrtc.IceCandidate] created based on this [IceCandidate].
   */
  fun intoWebRtc(): WIceCandidate {
    return WIceCandidate(sdpMid, sdpMLineIndex, sdp)
  }

  /** Converts this [IceCandidate] into a [Map] which can be returned to the Flutter side. */
  fun asFlutterResult(): Map<String, Any> =
      mapOf("sdpMid" to sdpMid, "sdpMLineIndex" to sdpMLineIndex, "candidate" to sdp)
}

File: ios/Classes/model/IceCandidate.swift
import WebRTC

/// Representation of an `RTCIceCandidate`.
class IceCandidate {
  /// mID of this `IceCandidate`.
  var sdpMid: String

  /// `sdpMLineIndex` of this `IceCandidate`.
  var sdpMLineIndex: Int

  /// `candidate` of this `IceCandidate`.
  var candidate: String

  /// Initializes a new `IceCandidate` based on the method call received from
  /// Flutter side.
  init(candidate: RTCIceCandidate) {
    self.sdpMid = candidate.sdpMid!
    self.candidate = candidate.sdp
    self.sdpMLineIndex = Int(candidate.sdpMLineIndex)
  }

  /// Initializes a new `IceCandidate` with the provided data.
  init(sdpMid: String, sdpMLineIndex: Int, candidate: String) {
    self.sdpMid = sdpMid
    self.sdpMLineIndex = sdpMLineIndex
    self.candidate = candidate
  }

  /// Converts this `IceCandidate` into an `RTCIceCandidate`.
  func intoWebRtc() -> RTCIceCandidate {
    RTCIceCandidate(
      sdp: self.candidate, sdpMLineIndex: Int32(self.sdpMLineIndex),
      sdpMid: self.sdpMid
    )
  }

  /// Converts this `IceCandidate` into a `Dictionary` which can be returned to
  /// Flutter side.
  func asFlutterResult() -> [String: Any] {
    [
      "sdpMid": self.sdpMid,
      "sdpMLineIndex": self.sdpMLineIndex,
      "candidate": self.candidate,
    ]
  }
}
